vllm/v1/worker/gpu_model_runner.py

```diff
@@ -2585,9 +2603,45 @@ def execute_model(
                 assert model_output_broadcast_data is not None
                 logits = model_output_broadcast_data["logits"]
 
-            # Apply structured output bitmasks if present
-            if scheduler_output.structured_output_request_ids:
-                apply_grammar_bitmask(scheduler_output, self.input_batch, logits)
+        self.execute_model_state = ExecuteModelState(
+            scheduler_output,
+            logits,
+            spec_decode_metadata,
+            spec_decode_common_attn_metadata,
+            hidden_states,
+            sample_hidden_states,
+            aux_hidden_states,
+            kv_connector_output,
+        )
+        return None
+
+    @torch.inference_mode
+    def sample_tokens(
+        self, grammar_output: "GrammarOutput | None"
```

Comment:
```suggestion
        return ExecuteModelState(
            scheduler_output,
            logits,
            spec_decode_metadata,
            spec_decode_common_attn_metadata,
            hidden_states,
            sample_hidden_states,
            aux_hidden_states,
            kv_connector_output,
        )

    @torch.inference_mode
    def sample_tokens(
        self, state: ExecuteModelState, grammar_output: "GrammarOutput | None"
```

Reply:
Discussed offline that we'll keep this as-is. Have added some comments to the methods in `WorkerBase` explaining that the two-stage design may change in future.